pilot/pkg/controllers/ipallocate/ipallocate.go

```diff
@@ -0,0 +1,402 @@
+	if se == nil {
+		return nil, nil
+	}
+	orig, err := json.Marshal(se)
+	if err != nil {
+		panic("failed to marshal original service entry")
+	}
+
+	if se.Status.String() == "" || forcedReassign {
+		// initialize status field
+		se.Status = v1alpha1.IstioStatus{}
+	} else {
+		foundAddresses := autoallocate.GetV2AddressesFromServiceEntry(se)
+		if len(foundAddresses) > 0 && !forcedReassign {
+
+			// this is likely a noop, but just to be safe we should check and potentially resolve conflict
+			for _, a := range foundAddresses {
+				c.markUsedOrQueueConflict(a, types.NamespacedName{
+					Name:      se.Name,
+					Namespace: se.Namespace,
+				})
+			}
+			return nil, nil // nothing to patch
+		}
+	}
+	addresses := c.nextAddresses(types.NamespacedName{
+		Name:      se.Name,
+		Namespace: se.Namespace,
+	})
+	// begin allocating new addresses
+	kludge := autoallocate.ConditionKludge(addresses)
+
+	se.Status.Conditions = append(se.Status.Conditions, &kludge)
+	modified, err := json.Marshal(se)
+	if err != nil {
+		panic("failed to marshal modified service entry")
+	}
+	return jsonpatch.CreateMergePatch(orig, modified)
+}
+
+type ipAllocator struct {
+	prefix netip.Prefix
+	used   map[netip.Addr]types.NamespacedName
+	next   netip.Addr
+}
+
+func newIPAllocator(p netip.Prefix) *ipAllocator {
+	n := p.Addr().Next()
+	return &ipAllocator{
+		prefix: p,
+		used:   map[netip.Addr]types.NamespacedName{},
+		next:   n,
+	}
+}
+
+func (i *ipAllocator) allocateNext(owner types.NamespacedName) (netip.Addr, error) {
+	n := i.next
+	var looped bool
+	if !n.IsValid() || !i.prefix.Contains(n) {
+		// we started outside of our range, that is unlucky but we can start over
+		looped = true
+		n = i.prefix.Addr().Next() // take the net address of the prefix and select next, this should be the first usable
+	}
+
+	for i.isUsed(n) {
+		n = n.Next()
+		if !n.IsValid() || !i.prefix.Contains(n) {
+			if looped {
+				// prefix is exhausted
+				return netip.Addr{}, fmt.Errorf("%s is exhausted", i.prefix.String())
+			}
+			looped = true
+			n = i.prefix.Addr().Next() // take the net address of the prefix and select next, this should be the first usable
+		}
+	}
+	i.markUsed(n, owner)
+
+	i.next = n.Next()
+
+	return n, nil
+}
+
+func (i ipAllocator) isUsed(n netip.Addr) bool {
+	_, found := i.used[n]
+	return found
+}
+
+func (i ipAllocator) isUsedBy(n netip.Addr, owner types.NamespacedName) (used, usedByOwner bool) {
+	if foundOwner, found := i.used[n]; found {
+		used = true // it is in use
+		res := strings.Compare(foundOwner.String(), owner.String())
+		usedByOwner = res == 0 // is it in use by the providec owner?
+		return
+	}
+	return
+}
+
+// markUsed will store the provided addr as used in this ipAllocator
+// markUsed returns true if the addr was already used and false if it was a new insert
+func (i ipAllocator) markUsed(n netip.Addr, owner types.NamespacedName) (used, usedByOwner bool) {
```

Comment:
I don't see the markUnused ( if a SE is deleted )

Reply:
correct. At this time we don't have logic that allows recycling IPs. That may be critical path or maybe we deem it OK to leave as a todo and work it out in a follow up. It's open for discussion.

Reply:
For IPv4 ( with the minuscule available ranges in large production systems with many clusters and pods ) - I think it's pretty critical. There are not so many /16 ranges sitting idle - and with multiple clusters (and external DNS) users will need to subdivide the range to each cluster (so we don't get different clusters assigning same IP, in particular with the incremental allocation).

For IPv6 - I mentioned in a different comment, I don't think incremental allocation is right ( for the GA), I don't mind it for the initial implementation but worth adding a comment that it's unlikely the best.

Reply:
I.e. - for Ipv6 we may not need markUnused if we switch to sha or random allocations, for IPv4 - I can't see how we'll manage the space between clusters and other uses.